functoria/lib/functoria.ml

```diff
@@ -250,3 +247,161 @@ module type KEY =
+          Codegen.newline fmt;
+          R.ok ())
+
+  let clean _i = Bos.OS.Path.delete file
+
+  let name = "key"
+
+end
+
+let keys (argv: argv impl) = impl @@ object
+    inherit base_configurable
+    method ty = job
+    method name = Keys.name
+    method module_name = Key.module_name
+    method !configure = Keys.configure
+    method !clean = Keys.clean
+    method !packages = Key.pure [package "functoria-runtime"]
+    method !deps = [ abstract argv ]
+    method !connect info modname = function
+      | [ argv ] ->
+        Fmt.strf
+          "return (Functoria_runtime.with_argv (List.map fst %s.runtime_keys) %S %s)"
+          modname (Info.name info) argv
+      | _ -> failwith "The keys connect should receive exactly one argument."
+  end
+
+(* Module emiting a file containing all the build information. *)
+
+type info = Info
+let info = Type Info
+
+let pp_libraries fmt l =
+  Fmt.pf fmt "[@ %a]"
+    Fmt.(iter ~sep:(unit ";@ ") List.iter @@ fmt "%S") l
+
+let pp_packages fmt l =
+  Fmt.pf fmt "[@ %a]"
+    Fmt.(iter ~sep:(unit ";@ ") List.iter @@
+         (fun fmt (n, v) -> pf fmt "%S, %S" n v)
+        ) l
+
+let pp_dump_pkgs module_name fmt (name, pkg, libs) =
+  Fmt.pf fmt
+    "%s.{@ name = %S;@ \
+     @[<v 2>packages = %a@]@ ;@ @[<v 2>libraries = %a@]@ }"
+    module_name name
+    pp_packages (String.Map.bindings pkg)
+    pp_libraries (String.Set.elements libs)
+
+let app_info ?(type_modname="Functoria_info")  ?(gen_modname="Info_gen") () =
+  impl @@ object
+    inherit base_configurable
+    method ty = info
+    method name = "info"
+    val file = Fpath.(v (String.Ascii.lowercase gen_modname) + "ml")
+    method module_name = gen_modname
+    method !packages = Key.pure [package "functoria-runtime"]
+    method !connect _ modname _ = Fmt.strf "return %s.info" modname
+
+    method !clean _i = Bos.OS.Path.delete file
+
+    method !configure _i = Ok ()
+
+    method !build i =
+      Log.info (fun m -> m "Generating: %a" Fpath.pp file);
+      (* this used to call 'opam list --rec ..', but that leads to
+         non-reproducibility, since this uses the opam CUDF solver which
+         drops some packages (which are in the repositories configured for the
+         switch), see https://github.com/mirage/functoria/pull/189 for further
+         discussion on this before changing the code below.  *)
+      let opam_deps pkgs =
+        let pkgs_str = String.concat ~sep:"," pkgs in
+        let cmd =
+          Bos.Cmd.(v "opam" % "list" % "--installed" % "-s"
+                   % "--color=never" % "--depopts" % "--resolve" % pkgs_str
+                   % "--columns" % "name,version")
+        in
+        (Bos.OS.Cmd.run_out cmd |> Bos.OS.Cmd.out_lines) >>= fun (deps, _) ->
+        let deps =
+          List.fold_left (fun acc s ->
+              match String.cuts ~empty:false ~sep:" " s with
+              | [n; v] -> (n, v) :: acc
+              | _ -> acc
```

Comment:
this looks a bit brittle, to just drop anything without a whitespace (or with multiple ones). if it can't happen, I'd be fine with an `assert false` here, if it can happen, maybe a warning message should be emitted?